### PR TITLE
feat(ops): indexing & slicing — slice/select/narrow + index_select (CPU + CUDA)

### LINKED ISSUE #10: [Feature] Indexing & slicing (CPU + CUDA)
# [Feature] Indexing & slicing (CPU + CUDA)

> **Type:** `[Feature]` — view-producing ops
> **Roadmap reference:** Eager Tensor Engine, README Phase 1 item 4
> **Blocks:** dataloader patterns, `nn::Embedding`, attention masking, gather/scatter style ops
> **Depends on:** Issue 02 (Tensor + view), Issue 03 (dispatch + parity harness)

---

## 1. Background

Issue 03 shipped element-wise ops that consume tensors of arbitrary shape, but there is still no first-class way to **address sub-regions** of a tensor. Slicing, single-row selection, and gather-style indexing are the bread-and-butter operations of any DL pipeline: dataloaders read mini-batches via `dataset[start:end]`, embedding tables are indexed by token id, attention masks are constructed from `arange` slices, and so on.

This issue ships the canonical view-producing ops so the rest of Phase 1 (and all of Phase 2/3) can write idiomatic indexing without resorting to manual stride math.

We deliberately split this from reductions (Issue 04): reductions consume strided inputs and produce values, while indexing produces **views** that share storage. Different testing concerns (alias safety, no-copy guarantees) and different kernels (no kernel at all on the cheap path).

---

### 2.1 Functional

- **F1.** Basic slice: `t.slice(int dim, int64_t start, int64_t end, int64_t step = 1)` returns a view sharing storage with `t`.
- **F2.** Single-element selection: `t.select(int dim, int64_t index)` returns a view with the selected dimension dropped.
- **F3.** Narrow: `t.narrow(int dim, int64_t start, int64_t length)` is sugar for `slice(dim, start, start + length, 1)`.
- **F4.** Index by an integer tensor (advanced indexing along a single axis): `Tensor index_select(const Tensor& src, int dim, const Tensor& indices)` returns a fresh contiguous tensor (not a view) of shape `src.shape with src.shape[dim] replaced by indices.numel()`.
- **F5.** Negative indices and negative slice bounds are normalised against `shape[dim]`. Out-of-range indices throw `ShapeError`.
- **F6.** Empty slices (`start == end`) produce a tensor with size 0 along that dim — valid, no error.
- **F7.** `slice`/`select`/`narrow` are zero-copy (no kernel launch). `index_select` is dispatched on the source device.
- **F8.** `slice` with `step > 1` is supported and reflected in the resulting stride; `step <= 0` throws `ShapeError`.

### 2.2 Non-functional

- **N1.** Numerical parity vs PyTorch reference: integer-exact for `index_select`; bit-exact for `slice`/`select`/`narrow` (they're pure pointer-arithmetic).
- **N2.** `slice`/`select`/`narrow` are O(1) in tensor numel — they only touch `shape`, `stride`, `offset`, never the data.
- **N3.** `index_select` on CUDA at 1M-element fp32 source + 1k-element index uses one kernel launch and lands within 25 % of a hand-written reference gather kernel.
- **N4.** All indexing ops listed in `docs/ops.md` with view/copy semantics noted explicitly.

---

## 3. Feature Description

```cpp
class Tensor {
public:
  // Existing methods unchanged. New zero-copy view methods:
  Tensor slice (int dim, int64_t start, int64_t end, int64_t step = 1) const;
  Tensor select(int dim, int64_t index) const;
  Tensor narrow(int dim, int64_t start, int64_t length) const;
};

namespace ctorch {
// Free function — produces a fresh contiguous output, dispatches on device.
Tensor index_select(const Tensor& src, int dim, const Tensor& indices);
}
```

End-to-end example:

```cpp
auto batch  = dataset.slice(/*dim=*/0, /*start=*/start, /*end=*/start + 32);
auto first  = batch.select(/*dim=*/0, /*index=*/0);
auto picked = ctorch::index_select(embedding_table, /*dim=*/0, token_ids);
```

---

### 4.1 View ops (`slice` / `select` / `narrow`)

All three live in `src/tensor.cpp`. Pure metadata mutation:

```cpp
Tensor Tensor::slice(int dim, int64_t start, int64_t end, int64_t step) const {
  // Normalise dim, start, end. Validate bounds.
  auto out = std::make_shared<detail::TensorImpl>();
  out->storage = impl_->storage;
  out->dt     = impl_->dt;
  out->shape  = impl_->shape;
  out->stride = impl_->stride;
  out->shape[dim]  = ceil_div(end - start, step);
  out->stride[dim] = impl_->stride[dim] * step;
  out->offset      = impl_->offset + start * impl_->stride[dim];
  return Tensor{out};
}

Tensor Tensor::select(int dim, int64_t idx) const {
  // Same idea, but drop dim from shape/stride.
}
```

`select` returning a tensor with `ndim - 1` is fine because our `Tensor` already supports rank-0 results (numel=1). The dispatch table doesn't need any changes.

### 4.2 `index_select` — front door

`include/ctorch/ops/indexing.h`:

```cpp
Tensor index_select(const Tensor& src, int dim, const Tensor& indices);
```

Front-door checks: `indices.dtype()` must be `int32` or `int64`; cross-device error if `src.device() != indices.device()`; out-of-range indices flagged in the kernel (or pre-checked on CPU).

The output shape is `src.shape with src.shape[dim] replaced by indices.numel()`. A fresh contiguous output tensor is allocated, then the kernel does the gather.

### 4.3 CPU gather kernel

```cpp
template <class T, class I>
void index_select_cpu(const Tensor& src, int dim, const Tensor& indices, Tensor& out) {
  // For each output position, decode coords; the dim-th coord is replaced by
  // indices[coord_at_dim]; load src[...] and write out[...].
}
```

Switch on `src.dtype()` outer, `indices.dtype()` inner (only `int32`/`int64`). One kernel template, two type axes.

### 4.4 CUDA gather kernel

Two paths:
- **Innermost dim**: each output element is one thread. Coalesced reads from `src` because `indices[i]` controls the slow stride only.
- **Non-innermost dim**: still one-element-per-thread but reads are non-coalesced. Acceptable for correctness; optimisation deferred.

Block size 256, grid `ceil_div(out.numel(), 256)`. Bounds-check: kernel emits an `int` flag in shared memory if any thread saw an out-of-range index, copied back to host post-launch and turned into `ShapeError`.

### 4.5 Op keys

```cpp
struct IndexSelectOp {
  using fn_t = void(*)(const Tensor& src, int dim, const Tensor& idx, Tensor& out);
};
```

Registered for both CPU and CUDA via the same `extern "C" ctorch_register_*` linkage trick from Issue 03 commit `d591a7f`.

### 4.6 Files added by this issue

```
include/ctorch/
└── ops/
    └── indexing.h            # public index_select declaration

src/
├── tensor.cpp                # extend with slice/select/narrow
└── ops/
    ├── indexing_cpu.cpp      # index_select_cpu + front-door + registrar
    └── indexing_cuda.cu      # index_select_cuda

include/ctorch/tensor.h       # extend with slice/select/narrow declarations

tests/tensor/
├── slice_test.cpp            # slice / select / narrow view semantics
└── view_alias_test.cpp       # round-trip: modify source ⇒ view sees change

tests/ops/
├── index_select_test.cpp     # functional: gather a few rows
└── index_select_dtype_test.cpp  # int32 vs int64 indices, dtype mismatch errors

tests/parity/
└── indexing_parity_test.cpp  # extends gen_parity.py
```

---

## 5. Acceptance Criteria

- [ ] `slice` / `select` / `narrow` produce views that share storage with the source (`.storage().data()` identical, `.use_count() > 1`).
- [ ] Negative dims and negative slice bounds are accepted and normalised.
- [ ] Step > 1 produces the right shape and stride; step <= 0 throws `ShapeError`.
- [ ] Empty slice (`start == end`) produces a valid 0-sized tensor; subsequent ops don't crash.
- [ ] `index_select` parity tests pass for `(int32 indices, int64 indices) × (float32, float64, int32, int64) × (innermost dim, non-innermost dim)`.
- [ ] Out-of-range indices raise `ShapeError` on both backends.
- [ ] CUDA `index_select` benchmark within 25 % of a hand-written reference gather on a 1M-element source / 1k-element index pair.
- [ ] `slice`/`select`/`narrow` are O(1) in tensor numel (verified via a microbenchmark over varying input sizes — wall-clock should stay flat).

---

## 6. Suggested commit breakdown

1. `feat(tensor): slice / select / narrow zero-copy view methods`
2. `test(tensor): slice/select/narrow view semantics + alias tests`
3. `feat(ops): public index_select front-door + op key`
4. `feat(ops): CPU index_select kernel with int32 / int64 indices`
5. `feat(ops): CUDA index_select with bounds-check flag`
6. `chore(test): commit indexing parity fixtures + extend gen_parity.py`
7. `bench: CUDA index_select vs reference gather`

---

## 7. Risks and trade-offs

| Risk | Mitigation |
| --- | --- |
| `slice` with negative step (reverse) is genuinely useful but doubles the stride-validation surface | Reject `step <= 0` for now; add reverse-slice as a follow-up |
| `index_select` bounds-check via shared-memory flag adds a sync point | The kernel is bound by global-memory traffic anyway; the extra sync is in the noise |
| Non-innermost-axis gather is slow vs a transposed version | Acceptable for correctness; revisit during the perf milestone |
| `select` produces a rank-0 tensor when the input was rank-1 — corner cases for future autograd | Lock in via test that `.numel() == 1` rank-0 tensors work in subsequent ops |
| Multi-dim advanced indexing (`t[i, j, :]`) is much more complex than single-axis `index_select` | Out of scope; documented as a planned follow-up |

---

## 8. Out of scope

- Multi-axis advanced indexing (`t[idx0, idx1, :]`) — separate `[Feature]` issue.
- `gather` / `scatter` (per-element general gather/scatter beyond a single dim) — separate issue.
- Boolean masking (`t[mask]`) — separate issue (requires `nonzero` first).
- Backward / autograd integration — Autograd milestone.
- `t[i:j] = value` (in-place index assignment) — separate `[Feature]` issue.

### PR DESCRIPTION
**Pull request by Hayden727**
## Summary
Implements [issue #10](https://github.com/Hayden727/ctorch/issues/10): zero-copy view-producing ops on `Tensor` and a front-door `index_select` that dispatches per device.

- **`Tensor::slice` / `select` / `narrow`** — pure metadata mutation, shared storage. PyTorch-style clamping for slice bounds; `ShapeError` for `step <= 0`, OOR `select` index, and out-of-range `narrow` ranges.
- **`ctorch::index_select`** — fresh contiguous output, int32/int64 indices, cross-device + dtype + rank-0 + non-1-D index validation in the front-door.
  - **CPU**: pre-resolves indices into source-axis offsets (one bounds check per index entry, not per output element); odometer-walks the contiguous output.
  - **CUDA**: two-kernel pattern — validation kernel raises a host-readable error flag via `atomicMax` and writes resolved offsets to a scratch buffer; gather kernel then consumes the scratch.
- **Parity** vs `np.take` across `f32 / f64 / i32 / i64` sources × `int32 / int64` indices × 1-D / 2-D / 3-D layouts (6 fixtures × 3 .npy files each).
- `docs/ops.md` and README roadmap updated.

## Test plan
- [x] `slice_test` (8 tests) — basic slice, step, negative bounds, dim normalisation, clamping, empty slice, bad step, rank-0 rejection
- [x] `view_alias_test` (4 tests) — bidirectional aliasing, chained views, lifetime
- [x] `index_select_test` (10 tests) — leading / inner / 3-D axes, int32 indices, negative indices, OOR rejection, strided source
- [x] `index_select_dtype_test` (7 tests) — dtype preservation, reject float / bool / non-1-D indices, dim OOR, rank-0 src, bfloat16
- [x] `indexing_parity_test` (6 cases) — exact equality vs `np.take`
- [x] All 233 tests pass on CPU build (was 190 baseline; +43 new tests)
- [ ] CUDA build verification (no toolchain on this dev machine; CI will exercise)

Closes #10

### CODE REVIEW

**Comment by Hayden727** [comment]
@codex review.

**Comment by Hayden727** [comment]
@codex review.

**Comment by Hayden727** [comment]
@codex review.

**Comment by Hayden727** [comment]
@codex review.
